File: apps/control-ui/src/api/selectiveImportWire.ts
import type {
	SelectiveImportApplyRequest as WireApplyRequest,
	SelectiveImportSelection as WireSelection,
} from "./generated/light-wire";
import type {
	SelectiveImportApplyRequest,
	SelectiveImportCatalog,
	SelectiveImportObjectKey,
	SelectiveImportOutcome,
	SelectiveImportPreview,
	SelectiveImportProfileKey,
	SelectiveImportSelection,
} from "./selectiveImportModels";
import {
	SELECTIVE_IMPORT_BLOCKER_TYPES,
	type SelectiveImportBlockerType,
} from "./selectiveImportModels";

export function selectiveImportSelectionToWire(
	selection: SelectiveImportSelection,
): WireSelection {
	return {
		selected_objects: selection.selectedObjects,
		conflict_resolutions: selection.conflictResolutions,
		profile_conflict_resolutions: selection.profileConflictResolutions.map(
			(choice) => ({
				key: {
					profile_id: choice.key.profileId,
					revision: choice.key.revision,
				},
				resolution: choice.resolution,
			}),
		),
	};
}

export function selectiveImportApplyToWire(
	request: SelectiveImportApplyRequest,
): WireApplyRequest {
	return {
		request_id: request.requestId,
		expected_source_revision: request.expectedSourceRevision,
		expected_target_revision: request.expectedTargetRevision,
		...selectiveImportSelectionToWire(request),
	};
}

export function selectiveImportCatalogFromWire(
	value: unknown,
): SelectiveImportCatalog {
	const body = record(value, "Selective Import catalog");
	return {
		sourceShowId: text(body.source_show_id, "source_show_id"),
		sourceShowName: text(body.source_show_name, "source_show_name"),
		sourceRevision: integer(body.source_revision, "source_revision"),
		objects: array(body.objects, "objects").map((entry, index) => {
			const object = record(entry, `objects[${index}]`);
			return {
				key: objectKey(object.key, `objects[${index}].key`),
				objectRevision: integer(
					object.object_revision,
					`objects[${index}].object_revision`,
				),
				displayName: text(
					object.display_name,
					`objects[${index}].display_name`,
				),
			};
		}),
	};
}

export function selectiveImportPreviewFromWire(
	value: unknown,
): SelectiveImportPreview {
	const body = record(value, "Selective Import preview");
	return {
		sourceShowId: text(body.source_show_id, "source_show_id"),
		targetShowId: text(body.target_show_id, "target_show_id"),
		sourceRevision: integer(body.source_revision, "source_revision"),
		targetRevision: integer(body.target_revision, "target_revision"),
		objects: array(body.objects, "objects").map(mapObjectPreview),
		dependencies: array(body.dependencies, "dependencies").map(mapDependency),
		conflicts: array(body.conflicts, "conflicts").map(mapConflict),
		profiles: array(body.profiles, "profiles").map(mapProfile),
		managedAssets: array(body.managed_assets, "managed_assets").map(mapAsset),
		blockers: array(body.blockers, "blockers").map(mapBlocker),
		canApply: flag(body.can_apply, "can_apply"),
	};
}

export function selectiveImportOutcomeFromWire(
	value: unknown,
): SelectiveImportOutcome {
	const body = record(value, "Selective Import outcome");
	return {
		requestId: text(body.request_id, "request_id"),
		correlationId: text(body.correlation_id, "correlation_id"),
		changed: flag(body.changed, "changed"),
		showId: text(body.show_id, "show_id"),
		showRevision: integer(body.show_revision, "show_revision"),
		eventSequence:
			body.event_sequence === null || body.event_sequence === undefined
				? null
				: integer(body.event_sequence, "event_sequence"),
		objectChanges: array(body.objects, "objects").map((entry, index) => {
			const object = record(entry, `objects[${index}]`);
			return {
				key: objectKey(object.key, `objects[${index}].key`),
				objectRevision: integer(
					object.object_revision,
					`objects[${index}].object_revision`,
				),
				body: object.body,
			};
		}),
		outcomes: array(body.outcomes, "outcomes").map(mapObjectPreview),
		profileChanges: array(body.profiles, "profiles").map((entry, index) => {
			const profile = record(entry, `profiles[${index}]`);
			return {
				source: profileKey(profile.source, `profiles[${index}].source`),
				destination: profileKey(
					profile.destination,
					`profiles[${index}].destination`,
				),
				digest: text(profile.digest, `profiles[${index}].digest`),
			};
		}),
		managedAssets: array(body.managed_assets, "managed_assets").map(
			(entry, index) => {
				const asset = record(entry, `managed_assets[${index}]`);
				return {
					assetId: text(asset.asset_id, `managed_assets[${index}].asset_id`),
					revision: integer(asset.revision, `managed_assets[${index}].revision`),
				};
			},
		),
	};
}

function mapObjectPreview(value: unknown, index: number) {
	const item = record(value, `objects[${index}]`);
	const action = record(item.action, `objects[${index}].action`);
	const actionType = oneOf(
		action.type,
		`objects[${index}].action.type`,
		[
			"import_preserving_id",
			"skip_identical",
			"keep_destination",
			"replace_destination",
			"duplicate",
			"blocked_conflict",
		] as const,
	);
	if (actionType === "duplicate") {
		objectKey(action.destination, `objects[${index}].action.destination`);
	}
	return {
		source: objectKey(item.source, `objects[${index}].source`),
		destination: objectKey(item.destination, `objects[${index}].destination`),
		action: actionType,
	};
}

function mapDependency(value: unknown, index: number) {
	const item = record(value, `dependencies[${index}]`);
	return {
		owner: objectKey(item.owner, `dependencies[${index}].owner`),
		dependency: objectKey(item.dependency, `dependencies[${index}].dependency`),
		disposition: oneOf(item.disposition, `dependencies[${index}].disposition`, [
			"selected",
			"included",
			"bound_to_destination",
			"missing",
		] as const),
	};
}

function mapConflict(value: unknown, index: number) {
	const item = record(value, `conflicts[${index}]`);
	return {
		key: objectKey(item.key, `conflicts[${index}].key`),
		resolution:
			item.resolution === null
				? null
				: oneOf(item.resolution, `conflicts[${index}].resolution`, [
						"keep_destination",
						"replace_destination",
						"duplicate",
					] as const),
	};
}

function mapProfile(value: unknown, index: number) {
	const item = record(value, `profiles[${index}]`);
	const action = record(item.action, `profiles[${index}].action`);
	const actionType = oneOf(action.type, `profiles[${index}].action.type`, [
		"copy",
		"skip_identical",
		"keep_destination",
		"duplicate",
		"blocked_conflict",
		"missing",
	] as const);
	if (actionType === "duplicate") {
		profileKey(action.destination, `profiles[${index}].action.destination`);
	}
	return {
		source: profileKey(item.source, `profiles[${index}].source`),
		destination: profileKey(item.destination, `profiles[${index}].destination`),
		action: actionType,
	};
}

function mapAsset(value: unknown, index: number) {
	const item = record(value, `managed_assets[${index}]`);
	const asset = record(item.asset, `managed_assets[${index}].asset`);
	return {
		asset: {
			assetId: text(asset.asset_id, `managed_assets[${index}].asset.asset_id`),
			revision: integer(asset.revision, `managed_assets[${index}].asset.revision`),
		},
		action: oneOf(item.action, `managed_assets[${index}].action`, [
			"copy",
			"skip_identical",
			"missing",
			"blocked_conflict",
		] as const),
	};
}

function mapBlocker(value: unknown, index: number) {
	const item = record(value, `blockers[${index}]`);
	const type = oneOf(
		item.type,
		`blockers[${index}].type`,
		SELECTIVE_IMPORT_BLOCKER_TYPES,
	);
	return {
		type,
		summary: [label(type), blockerDetail(type, item, index)].filter(Boolean).join(": "),
	};
}

function blockerDetail(
	type: SelectiveImportBlockerType,
	item: Record<string, unknown>,
	index: number,
) {
	const path = `blockers[${index}]`;
	switch (type) {
		case "empty_selection":
		case "same_show":
			return null;
		case "unsupported_object":
		case "object_conflict":
			return objectIdentity(item.key, `${path}.key`);
		case "missing_object":
			objectKeyOrNull(item.required_by, `${path}.required_by`);
			return objectIdentity(item.key, `${path}.key`);
		case "invalid_resolution":
		case "invalid_descriptor":
			objectKey(item.key, `${path}.key`);
			return text(item.message, `${path}.message`);
		case "invalid_profile_resolution":
			profileKey(item.key, `${path}.key`);
			return text(item.message, `${path}.message`);
		case "missing_profile":
			objectKey(item.required_by, `${path}.required_by`);
			return profileIdentity(item.key, `${path}.key`);
		case "profile_conflict":
			return profileIdentity(item.key, `${path}.key`);
		case "missing_managed_asset":
		case "managed_asset_conflict":
			return assetIdentity(item.asset, `${path}.asset`);
		case "reference_rewrite":
			objectKey(item.owner, `${path}.owner`);
			return text(item.message, `${path}.message`);
		case "candidate_invalid":
			return text(item.message, `${path}.message`);
	}
}

function objectKey(value: unknown, name: string): SelectiveImportObjectKey {
	const key = record(value, name);
	return { kind: text(key.kind, `${name}.kind`), id: text(key.id, `${name}.id`) };
}

function objectKeyOrNull(value: unknown, name: string) {
	if (value !== null) objectKey(value, name);
}

function objectIdentity(value: unknown, name: string) {
	const key = objectKey(value, name);
	return `${key.kind}/${key.id}`;
}

function profileKey(value: unknown, name: string): SelectiveImportProfileKey {
	const key = record(value, name);
	return {
		profileId: text(key.profile_id, `${name}.profile_id`),
		revision: integer(key.revision, `${name}.revision`),
	};
}

function profileIdentity(value: unknown, name: string) {
	const key = profileKey(value, name);
	return `${key.profileId} Revision ${key.revision}`;
}

function assetIdentity(value: unknown, name: string) {
	const asset = record(value, name);
	const id = text(asset.asset_id, `${name}.asset_id`);
	const revision = integer(asset.revision, `${name}.revision`);
	return `${id} Revision ${revision}`;
}

function record(value: unknown, name: string): Record<string, unknown> {
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error(`${name} is not an object`);
	}
	return value as Record<string, unknown>;
}

function array(value: unknown, name: string): unknown[] {
	if (!Array.isArray(value)) throw new Error(`${name} is not an array`);
	return value;
}

function text(value: unknown, name: string): string {
	if (typeof value !== "string") throw new Error(`${name} is not a string`);
	return value;
}

function integer(value: unknown, name: string): number {
	if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) {
		throw new Error(`${name} is not a non-negative safe integer`);
	}
	return value;
}

function flag(value: unknown, name: string): boolean {
	if (typeof value !== "boolean") throw new Error(`${name} is not a boolean`);
	return value;
}

function oneOf<const T extends readonly string[]>(
	value: unknown,
	name: string,
	allowed: T,
): T[number] {
	const candidate = text(value, name);
	if (!allowed.includes(candidate)) {
		throw new Error(`${name} has an unsupported value`);
	}
	return candidate as T[number];
}

function label(value: string) {
	return value.replaceAll("_", " ").replace(/\b\w/g, (letter) => letter.toUpperCase());
}

File: crates/server/src/runtime/selective_import_wire.rs
//! Lossless conversion between v2 selective-import DTOs and application models.

use light_application as app;
use light_core::{FixtureId, ShowId};
use light_show::{PortableShowDocument, PortableShowObjectKey, PortableShowRevision};
use light_wire::v2::selective_import as wire;
use std::collections::{BTreeMap, BTreeSet};

pub(super) fn application_request(
    source_show_id: ShowId,
    target_show_id: ShowId,
    selection: wire::SelectiveImportSelection,
) -> Result<app::SelectiveShowImportRequest, String> {
    if selection.selected_objects.len() > 4_096
        || selection.conflict_resolutions.len() > 4_096
        || selection.profile_conflict_resolutions.len() > 4_096
    {
        return Err("selective import requests are limited to 4096 entries per section".into());
    }
    let selected_objects = unique_object_keys(selection.selected_objects, "selected object")?;
    let mut conflict_resolutions = BTreeMap::new();
    for choice in selection.conflict_resolutions {
        let key = application_key(choice.key)?;
        if conflict_resolutions
            .insert(key, conflict_resolution(choice.resolution))
            .is_some()
        {
            return Err("an object conflict has more than one resolution".into());
        }
    }
    let mut profile_conflict_resolutions = BTreeMap::new();
    for choice in selection.profile_conflict_resolutions {
        let key = application_profile_key(choice.key);
        if profile_conflict_resolutions
            .insert(key, profile_resolution(choice.resolution))
            .is_some()
        {
            return Err("a profile conflict has more than one resolution".into());
        }
    }
    Ok(app::SelectiveShowImportRequest {
        source_show_id,
        target_show_id,
        selected_objects,
        conflict_resolutions,
        profile_conflict_resolutions,
    })
}

pub(super) fn catalog(document: &PortableShowDocument) -> wire::SelectiveImportCatalog {
    wire::SelectiveImportCatalog {
        source_show_id: document.id().0,
        source_show_name: document.name().into(),
        source_revision: document.revision().value(),
        objects: document
            .objects()
            .map(|object| wire::SelectiveImportCatalogObject {
                key: wire_key(object.key()),
                object_revision: object.revision(),
                display_name: display_name(object.key(), object.body()),
            })
            .collect(),
    }
}

pub(super) fn preview(preview: app::SelectiveShowImportPreview) -> wire::SelectiveImportPreview {
    let can_apply = preview.can_apply();
    wire::SelectiveImportPreview {
        source_show_id: preview.request.source_show_id.0,
        target_show_id: preview.request.target_show_id.0,
        source_revision: preview.source_revision.value(),
        target_revision: preview.target_revision.value(),
        objects: preview.objects.into_iter().map(object_preview).collect(),
        dependencies: preview
            .dependencies
            .into_iter()
            .map(|dependency| wire::SelectiveImportDependency {
                owner: wire_key(&dependency.owner),
                dependency: wire_key(&dependency.dependency),
                disposition: dependency_disposition(dependency.disposition),
            })
            .collect(),
        conflicts: preview
            .conflicts
            .into_iter()
            .map(|conflict| wire::SelectiveImportConflict {
                key: wire_key(&conflict.key),
                resolution: conflict.resolution.map(wire_conflict_resolution),
            })
            .collect(),
        profiles: preview.profiles.into_iter().map(profile_preview).collect(),
        managed_assets: preview
            .managed_assets
            .into_iter()
            .map(|asset| wire::SelectiveImportManagedAssetPreview {
                asset: wire_asset(asset.asset),
                action: managed_asset_action(asset.action),
            })
            .collect(),
        blockers: preview.blockers.into_iter().map(blocker).collect(),
        can_apply,
    }
}

pub(super) fn outcome(result: app::SelectiveShowImportResult) -> wire::SelectiveImportOutcome {
    wire::SelectiveImportOutcome {
        request_id: result.context.request_id.clone().unwrap_or_default(),
        correlation_id: result.context.correlation_id,
        changed: result.changed,
        show_id: result.change.show_id.0,
        show_revision: result.change.show_revision.value(),
        event_sequence: result.event_sequence,
        outcomes: result
            .change
            .outcomes
            .into_iter()
            .map(object_preview)
            .collect(),
        objects: result
            .change
            .objects
            .into_iter()
            .map(|object| wire::SelectiveImportOutcomeObjectChange {
                key: wire_key(&object.key),
                object_revision: object.object_revision,
                body: object.body,
            })
            .collect(),
        profiles: result
            .change
            .profiles
            .into_iter()
            .map(|profile| wire::SelectiveImportProfileChange {
                source: wire_profile_key(profile.source),
                destination: wire_profile_key(profile.destination),
                digest: profile.digest,
            })
            .collect(),
        managed_assets: result
            .change
            .managed_assets
            .into_iter()
            .map(wire_asset)
            .collect(),
    }
}

pub(super) fn expected_revision(value: u64) -> PortableShowRevision {
    PortableShowRevision::from_value(value)
}

fn unique_object_keys(
    keys: Vec<wire::SelectiveImportObjectKey>,
    label: &str,
) -> Result<BTreeSet<PortableShowObjectKey>, String> {
    let mut output = BTreeSet::new();
    for key in keys {
        if !output.insert(application_key(key)?) {
            return Err(format!("duplicate {label}"));
        }
    }
    Ok(output)
}

fn application_key(key: wire::SelectiveImportObjectKey) -> Result<PortableShowObjectKey, String> {
    for (label, value) in [("object kind", &key.kind), ("object id", &key.id)] {
        if value.is_empty() || value.len() > 256 || value.chars().any(char::is_control) {
            return Err(format!("{label} must contain 1-256 printable characters"));
        }
    }
    Ok(PortableShowObjectKey::new(key.kind, key.id))
}

fn application_profile_key(key: wire::SelectiveImportProfileKey) -> app::ImportProfileKey {
    app::ImportProfileKey {
        profile_id: FixtureId(key.profile_id),
        revision: key.revision,
    }
}

fn conflict_resolution(
    value: wire::SelectiveImportConflictResolution,
) -> app::ImportConflictResolution {
    match value {
        wire::SelectiveImportConflictResolution::KeepDestination => {
            app::ImportConflictResolution::KeepDestination
        }
        wire::SelectiveImportConflictResolution::ReplaceDestination => {
            app::ImportConflictResolution::ReplaceDestination
        }
        wire::SelectiveImportConflictResolution::Duplicate => {
            app::ImportConflictResolution::Duplicate
        }
    }
}

fn profile_resolution(
    value: wire::SelectiveImportProfileConflictResolution,
) -> app::ImportProfileConflictResolution {
    match value {
        wire::SelectiveImportProfileConflictResolution::KeepDestination => {
            app::ImportProfileConflictResolution::KeepDestination
        }
        wire::SelectiveImportProfileConflictResolution::Duplicate => {
            app::ImportProfileConflictResolution::Duplicate
        }
    }
}

fn wire_conflict_resolution(
    value: app::ImportConflictResolution,
) -> wire::SelectiveImportConflictResolution {
    match value {
        app::ImportConflictResolution::KeepDestination => {
            wire::SelectiveImportConflictResolution::KeepDestination
        }
        app::ImportConflictResolution::ReplaceDestination => {
            wire::SelectiveImportConflictResolution::ReplaceDestination
        }
        app::ImportConflictResolution::Duplicate => {
            wire::SelectiveImportConflictResolution::Duplicate
        }
    }
}

fn object_preview(value: app::AppliedImportObject) -> wire::SelectiveImportObjectPreview {
    let action = match value.action {
        app::ImportObjectAction::ImportPreservingId => {
            wire::SelectiveImportObjectAction::ImportPreservingId
        }
        app::ImportObjectAction::SkipIdentical => wire::SelectiveImportObjectAction::SkipIdentical,
        app::ImportObjectAction::KeepDestination => {
            wire::SelectiveImportObjectAction::KeepDestination
        }
        app::ImportObjectAction::ReplaceDestination => {
            wire::SelectiveImportObjectAction::ReplaceDestination
        }
        app::ImportObjectAction::Duplicate { destination } => {
            wire::SelectiveImportObjectAction::Duplicate {
                destination: wire_key(&destination),
            }
        }
        app::ImportObjectAction::BlockedConflict => {
            wire::SelectiveImportObjectAction::BlockedConflict
        }
    };
    wire::SelectiveImportObjectPreview {
        source: wire_key(&value.source),
        destination: wire_key(&value.destination),
        action,
    }
}

fn dependency_disposition(
    value: app::ImportDependencyDisposition,
) -> wire::SelectiveImportDependencyDisposition {
    match value {
        app::ImportDependencyDisposition::Selected => {
            wire::SelectiveImportDependencyDisposition::Selected
        }
        app::ImportDependencyDisposition::Included => {
            wire::SelectiveImportDependencyDisposition::Included
        }
        app::ImportDependencyDisposition::BoundToDestination => {
            wire::SelectiveImportDependencyDisposition::BoundToDestination
        }
        app::ImportDependencyDisposition::Missing => {
            wire::SelectiveImportDependencyDisposition::Missing
        }
    }
}

fn profile_preview(value: app::ImportProfilePreview) -> wire::SelectiveImportProfilePreview {
    let action = match value.action {
        app::ImportProfileAction::Copy => wire::SelectiveImportProfileAction::Copy,
        app::ImportProfileAction::SkipIdentical => {
            wire::SelectiveImportProfileAction::SkipIdentical
        }
        app::ImportProfileAction::KeepDestination => {
            wire::SelectiveImportProfileAction::KeepDestination
        }
        app::ImportProfileAction::Duplicate { destination } => {
            wire::SelectiveImportProfileAction::Duplicate {
                destination: wire_profile_key(destination),
            }
        }
        app::ImportProfileAction::BlockedConflict => {
            wire::SelectiveImportProfileAction::BlockedConflict
        }
        app::ImportProfileAction::Missing => wire::SelectiveImportProfileAction::Missing,
    };
    wire::SelectiveImportProfilePreview {
        source: wire_profile_key(value.source),
        destination: wire_profile_key(value.destination),
        action,
    }
}

fn managed_asset_action(
    value: app::ImportManagedAssetAction,
) -> wire::SelectiveImportManagedAssetAction {
    match value {
        app::ImportManagedAssetAction::Copy => wire::SelectiveImportManagedAssetAction::Copy,
        app::ImportManagedAssetAction::SkipIdentical => {
            wire::SelectiveImportManagedAssetAction::SkipIdentical
        }
        app::ImportManagedAssetAction::Missing => wire::SelectiveImportManagedAssetAction::Missing,
        app::ImportManagedAssetAction::BlockedConflict => {
            wire::SelectiveImportManagedAssetAction::BlockedConflict
        }
    }
}

fn blocker(value: app::ImportBlocker) -> wire::SelectiveImportBlocker {
    use app::ImportBlocker as A;
    use wire::SelectiveImportBlocker as W;
    match value {
        A::EmptySelection => W::EmptySelection,
        A::SameShow => W::SameShow,
        A::UnsupportedObject { key } => W::UnsupportedObject {
            key: wire_key(&key),
        },
        A::MissingObject { key, required_by } => W::MissingObject {
            key: wire_key(&key),
            required_by: required_by.as_ref().map(wire_key),
        },
        A::ObjectConflict { key } => W::ObjectConflict {
            key: wire_key(&key),
        },
        A::InvalidResolution { key, message } => W::InvalidResolution {
            key: wire_key(&key),
            message,
        },
        A::InvalidProfileResolution { key, message } => W::InvalidProfileResolution {
            key: wire_profile_key(key),
            message,
        },
        A::InvalidDescriptor { key, message } => W::InvalidDescriptor {
            key: wire_key(&key),
            message,
        },
        A::MissingProfile { key, required_by } => W::MissingProfile {
            key: wire_profile_key(key),
            required_by: wire_key(&required_by),
        },
        A::ProfileConflict { key } => W::ProfileConflict {
            key: wire_profile_key(key),
        },
        A::MissingManagedAsset { asset } => W::MissingManagedAsset {
            asset: wire_asset(asset),
        },
        A::ManagedAssetConflict { asset } => W::ManagedAssetConflict {
            asset: wire_asset(asset),
        },
        A::ReferenceRewrite { owner, message } => W::ReferenceRewrite {
            owner: wire_key(&owner),
            message,
        },
        A::CandidateInvalid { message } => W::CandidateInvalid { message },
    }
}

fn wire_key(value: &PortableShowObjectKey) -> wire::SelectiveImportObjectKey {
    wire::SelectiveImportObjectKey {
        kind: value.kind().into(),
        id: value.id().into(),
    }
}

fn wire_profile_key(value: app::ImportProfileKey) -> wire::SelectiveImportProfileKey {
    wire::SelectiveImportProfileKey {
        profile_id: value.profile_id.0,
        revision: value.revision,
    }
}

fn wire_asset(value: app::AssetReference) -> wire::SelectiveImportAssetReference {
    wire::SelectiveImportAssetReference {
        asset_id: value.id.0,
        revision: value.revision.0,
    }
}

fn display_name(key: &PortableShowObjectKey, body: &serde_json::Value) -> String {
    for field in ["name", "label", "title", "number"] {
        if let Some(value) = body.get(field) {
            if let Some(text) = value.as_str().filter(|text| !text.trim().is_empty()) {
                return text.into();
            }
            if value.is_number() {
                return value.to_string();
            }
        }
    }
    key.id().into()
}
